### PR TITLE
Use NumPy C API to perform `DimShuffle` steps in its C implementation

### LINKED ISSUE #699: `DimShuffle`'s C code does not correctly handle `ndarray` views
Here's a simple example:
```python
import aesara.tensor as at


at.broadcast_to(0, (2, 3))[None].eval()
# array([[[  0, 122, -76],
#         [-88,  49,  86]]], dtype=int8)
```

Here's the same result, just computed more directly via the generated C thunk:
```python
import numpy as np
import aesara


x_at = at.vector()
x_fn = aesara.function([x_at], x_at[None])

x_fn.fn.fgraph.toposort()
# [InplaceDimShuffle{x,0}(<TensorType(float64, vector)>),
#  DeepCopyOp(InplaceDimShuffle{x,0}.0)]

dimshuffle_thunk = x_fn.fn.thunks[0]

dimshuffle_thunk
# <function aesara.graph.op.COp.make_c_thunk.<locals>.rval()>

dimshuffle_thunk.inputs[0][0] = np.broadcast_to(0, (6,))
dimshuffle_thunk()
dimshuffle_thunk.outputs[0][0]
# array([[              0,               0, 140009315547216,
#                      33,               0,               0]])

np.shares_memory(dimshuffle_thunk.inputs[0][0], dimshuffle_thunk.outputs[0][0])
# True
```

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR replaces the custom view array creation in `DimShuffle`'s C implementation with calls to NumPy's C API that match the Python implementation.  It also replaces the manual shape creation steps in `Reshape`'s C implementation with a call to a NumPy helper function.

Closes #699

### CODE REVIEW

File: aesara/tensor/c_code/dimshuffle.c [hunk 1]
```diff
@@ -1,104 +1,86 @@
...
-    if (params->inplace) {
-        basename = input;
-        Py_INCREF((PyObject*)basename);
-    } else {
-        basename =
-            (PyArrayObject*)PyArray_FromAny((PyObject*)input,
-                                            NULL, 0, 0, NPY_ARRAY_ALIGNED|NPY_ARRAY_ENSURECOPY, NULL);
-    }
+  if (!PyArray_IS_C_CONTIGUOUS(params->augment)) {
+    PyErr_SetString(PyExc_RuntimeError, "DimShuffle: param input_broadcastable must be C-contiguous.");
```
Comment (ricardoV94) [inline aesara/tensor/c_code/dimshuffle.c:11]:
```suggestion
    PyErr_SetString(PyExc_RuntimeError, "DimShuffle: param augment must be C-contiguous.");
```

Comment (ricardoV94) [reply aesara/tensor/c_code/dimshuffle.c:11]: Why must it be C-contiguous? The loop below should work fine regardless of layout no?

File: aesara/tensor/shape.py [hunk 2]
```diff
@@ -669,44 +669,32 @@ def infer_shape(self, fgraph, node, ishapes):
...
-            sdtype = node.inputs[1].type.dtype_specs()[1]
-            fail = sub["fail"]
-            params = sub["params"]
-            return (
-                """
-            assert (PyArray_NDIM(%(shp)s) == 1);
-            npy_intp new_dims[%(params)s->ndim];
-            PyArray_Dims newshape;
-            newshape.ptr = new_dims;
-            newshape.len = %(params)s->ndim;
```
Comment (ricardoV94) [inline aesara/tensor/shape.py:687]: Seems like the new code is not using the `ndim` anymore, which the failing test depended on: https://github.com/aesara-devs/aesara/runs/4530278695?check_suite_focus=true#step:6:336

https://github.com/Theano/Theano/pull/5616/files#r102949026

File: aesara/tensor/c_code/dimshuffle.c (hunk 1 above)
Comment (brandonwillard) [reply aesara/tensor/c_code/dimshuffle.c]: Not sure; those checks might've originally been there for the `PyArray_DATA` calls.  Regardless, we don't actually need those anymore&mdash;e.g. we could use `PyArray_GetPtr` in the new code instead.

Comment (brandonwillard) [reply aesara/tensor/c_code/dimshuffle.c]: I just made that change.

**Review by ricardoV94 — APPROVED** [review]

**Pull request merged**